### ISSUE TITLE
Ability to specify AWS service in use // General IAM overhaul

### ISSUE THREAD

**Issue by Miserlou** [Miserlou/Zappa#524]
IAM needs a serious overhaul. In addition to fixing #479 , I'd to replace the current 'default everything' policy with something much more specific. We could also have per-project policies and generally just avoid * wherever possible.

Rather than doing Chalice's automagic policy detection, I'd greatly prefer to have something explicit, something like:

```json
aws_services: ['s3', 'lambda', 'apigateway', 'ses', 'sns']
```

But, that may still be too broad. Input requested!

**Comment by mcrowson** [comment]
I think the automagic policy detection approach is less of a learning curve for people to new to Zappa.

**Comment by Miserlou** [comment]
If anybody wants to scoop that out of Chalice and modularize it, please be my guest!

**Comment by mathom** [comment]
It would be pretty simple to just let people list the AWS managed policies in there and attach em (for example "AmazonS3ReadOnlyAccess" is one). To take a look at the list they have try `aws iam list-policies --scope AWS --query 'Policies[].PolicyName'`

**Comment by pjz** [comment]
Maybe both?  If it gets scooped out of Chalice, have a magic `autodetect` value that's the default, or can be overridden. So either:

```
{
        "aws_services": ["s3", "lambda", "apigateway", "ses", "sns"]
}
```
or
```
{
        "aws_services": "autodetect"
}
```
would be valid, with the latter using some Chalice-like detection, or the former allowing it to be specified.  This would satisfy ease-of-use in the default case but also allow explicit specification for edge cases.

**Comment by mihow** [comment]
I agree that autodetecting would be pretty nice for getting started in a more secure way. I like @pjz 's suggestion of allowing muliple methods. What about a 3rd option that allows you specify a full IAM policy? Perhaps a path to a `project_role_iam_policy.json` file. That's similar to the way that Terraform does it. 

Something like:

`zappa generate_iam_policy > project_iam_policy.json`

Customize...

```
{ 
  "role_name": "MyLambdaRole",
  "role_policy": "project_iam_policy.json"
}
````

This looks like the meat of Chalice's auto-detection: 
https://github.com/awslabs/chalice/blob/3c5eed564b53b16e069eb179baca6845fc9554f1/chalice/analyzer.py

**Comment by Miserlou** [comment]
You can also provide that using the

        "assume_policy": "my_assume_policy.json", // optional, IAM
assume policy JSON file
        "attach_policy": "my_attach_policy.json", // optional, IAM
attach policy JSON file


https://github.com/Miserlou/Zappa#using-custom-aws-iam-roles-and-policies

**Comment by mihow** [comment]
doi! thanks @Miserlou

**Comment by Miserlou** [comment]
Of course, the documentation around that could be improved as part of this effort. 😸

**Comment by atif1996** [comment]
Interesting discussion.  Having automatic policies is a good place to start, but in a corporate environment, you will usually end up specifying them individually.  In many cases, the developer wont have sufficient privileges to allow zappa to generate the policy.  Instead, an admin will create the policies and specify them.

I'd suggest having something where you can either :
- Specify your roles (one for ApiGateway, and one for Lambda)
- Allow it to auto-generate, or
- Echo to the console that permissions are recommended.  

That way, the user has something to email the Admin when things don't work.

When going the auto-generated route though, I would imagine the developer would know what services they will use.  A place to simple list of the services you will use seems reasonable.

**Comment by Miserlou** [comment]
Bumping this because it's coming up in related CI/CD discussions.

For what it's worth, I tried scooping out the chalice policy detector and it simply doesn't work as advertised. It's useless.
